crates/common/src/result.rs

```diff
@@ -1,51 +1,64 @@
     #[serde(with = "serde_bytes")]
-    Deserialize(Vec<u8>),
+    Deserialize(Box<[u8]>),
     /// Something failed to serialize.
-    /// This should be rare or impossible for basically everything that implements Serialize.
+    /// This should be rare or impossible for almost everything that implements `Serialize`.
     Serialize(SerializedBytesError),
     /// Somehow we errored while erroring.
     /// For example, maybe we failed to serialize an error while attempting to serialize an error.
     ErrorWhileError,
     /// Something went wrong while writing or reading bytes to/from wasm memory.
-    /// this means something like "reading 16 bytes did not produce 2x WasmSize ints"
-    /// or maybe even "failed to write a byte to some pre-allocated wasm memory"
-    /// whatever this is it is very bad and probably not recoverable
+    /// Whatever this is it is very bad and probably not recoverable.
     Memory,
-    /// Failed to take bytes out of the guest and do something with it.
-    /// The string is whatever error message comes back from the interal process.
+    /// Host failed to take bytes out of the guest and do something with it.
+    /// The string is whatever error message comes back from the internal process.
     GuestResultHandling(String),
-    /// Something to do with guest logic that we don't know about
+    /// Error with guest logic that the host doesn't know about.
     Guest(String),
-    /// Something to do with host logic that we don't know about
+    /// Error with host logic that the guest doesn't know about.
     Host(String),
-    /// Something to do with host logic that we don't know about
+    /// Something to do with host logic that the guest doesn't know about
     /// AND wasm execution MUST immediately halt.
-    /// The Vec<u8> holds the encoded data as though the guest had returned.
+    /// The `Vec<u8>` holds the encoded data as though the guest had returned.
     HostShortCircuit(Vec<u8>),
-    /// Somehow wasmer failed to compile machine code from wasm byte code
+    /// Wasmer failed to compile machine code from wasm byte code.
     Compile(String),
-
+    /// The host failed to call a function in the guest.
     CallError(String),
-
+    /// Host attempted to interact with the module cache before it was initialized.
     UninitializedSerializedModuleCache,
 }
 
+impl WasmErrorInner {
+    /// Some errors indicate the wasm guest is potentially corrupt and so the
+    /// host MUST NOT reuse it (e.g. in a cache of wasm instances). Other errors
+    /// MAY NOT invalidate an instance cache on the host.
+    pub fn maybe_corrupt(&self) -> bool {
+        match self {
+            Self::PointerMap
+            | Self::ErrorWhileError
+            | Self::Memory
+            | Self::GuestResultHandling(_)
+            | Self::Compile(_)
+            | Self::CallError(_)
+            | Self::UninitializedSerializedModuleCache => true,
+            _ => false,
```

Comment:
consider filling out the false case with a bunch of ors (instead of the catchall) so that adding a new variant causes the compiler to force you to decide if it is a corruption related error or not.